### PR TITLE
Remove usage of std::function in MVI

### PR DESCRIPTION
**Pull request by bjjones**
Removes usage of a callback using std::function in favor of a C-style function pointer callback.

### CODE REVIEW

**Comment by bjjones** [comment]
This hasn't been tested. What is the best way to test the MVI? Building the WIP DML Backend patch right now to try testing with the full thing.

I'm a little unsure if I got the ComPtr operations correct - it's been awhile.

File: src/mvi/gpgmm_d3d12.cpp
Comment (bjjones) [inline src/mvi/gpgmm_d3d12.cpp:79]: Could stuff these params into a struct to clean it up a bit.

File: src/mvi/gpgmm_d3d12.cpp
Comment (bbernhar) [inline src/mvi/gpgmm_d3d12.cpp:76]: Any changes to this (exported) function also need to be made in `include/gpgmm_d3d12.h` or it won't compile.

File: src/mvi/gpgmm_d3d12.h
Comment (bbernhar) [inline src/mvi/gpgmm_d3d12.h:47]: This needs to be in `include/gpgmm_d3d12.h` or full GMM won't see it.

File: src/mvi/gpgmm_d3d12.cpp
Comment (bbernhar) [inline src/mvi/gpgmm_d3d12.cpp:381]: Remove? Doesn't seem to do anything.

**Comment by bbernhar** [comment]
> This hasn't been tested. What is the best way to test the MVI?

Probably need to come up with a end2end test specifically for the MVI (really basic), and perhaps re-use it for the Chromium CL.

**Comment by bbernhar** [comment]
Approach LGTM + left some comments.

File: src/mvi/gpgmm_d3d12.cpp
Comment (bbernhar) [reply src/mvi/gpgmm_d3d12.cpp:76]: Yup!

File: src/mvi/gpgmm_d3d12.h
Comment (bbernhar) [inline src/mvi/gpgmm_d3d12.h:74]: CreateHeap needs to also work with non-resources so passing resource-specific parameters like this won't work for other heap types. The call-back can only accept ID3D12Pageable as output which is what `CreateCommittedResourceCallback` should return and pass to CreateHeap.

File: src/mvi/gpgmm_d3d12.cpp
Comment (bjjones) [reply src/mvi/gpgmm_d3d12.cpp:381]: Ah ignore this for now. Made a mistake when rebasing.

File: src/mvi/gpgmm_d3d12.h
Comment (bjjones) [reply src/mvi/gpgmm_d3d12.h:74]: Any idea how to do that? Obviously I have to call CreateCommittedResource in a callback, which requires those resource-specific parameters. I have to get those to the callback somehow - but if I can't pass as parameters the only other option I can think of is some kind of binding model. 

Is it not an option to just pass a HeapParams struct that would contain a subset of possible heap creation parameters?

Comment (bjjones) [reply src/mvi/gpgmm_d3d12.h:74]: (Sorry, didn't see the edit)

Comment (bbernhar) [reply src/mvi/gpgmm_d3d12.h:74]: D3D12 needs a lot of HeapParams (for. CreateDescriptorHeap, CreateQueryHeap, CreateFence, etc).

I was wondering if we could create `struct CreateResourceWrapper` which gets initialized to the HeapParams specific to the type. Then pass `CreateResourceWrapper::CreateFn` to CreateHeap which calls the D3D creation method and returns the ID3D12Pageable provided by CreateHeap.

Comment (bjjones) [reply src/mvi/gpgmm_d3d12.h:74]: So `struct CreateResourceWrapper` would be a base and we'd create more structs that inherit from it i.e. `struct CreateCommittedResourceParams : CreateResourceWrapper` then we'd pass it through `CreateHeap`, pass it to the `CreateHeapFn`, then cast it to the appropriate type within `CreateHeapFn`?

Comment (bjjones) [reply src/mvi/gpgmm_d3d12.h]: Pushed a new draft for the full gpgmm, which I think implements the idea described. Seems like it will work but this method means we can't pass in our CreateHeapFn in the exported API. It doesn't seem like there's a workaround unless I've misinterpreted the design suggestion. How much of a problem is this?

Comment (bjjones) [reply src/mvi/gpgmm_d3d12.h]: It doesn't appear that the WebNN DML prototype actually calls the CreateHeap export...so we could just surround it with a NON_CHROMIUM_COMPLIANT ifdef and leave CreateHeapFn-based functionality - whereas the internal calls to CreateHeap could use the method from the current PR draft.

Comment (bjjones) [reply src/mvi/gpgmm_d3d12.h]: Pushed another draft. Needs polish but lmk if the approach is as expected.

File: src/gpgmm/d3d12/ResourceAllocatorD3D12.cpp
Comment (bjjones) [reply src/gpgmm/d3d12/ResourceAllocatorD3D12.cpp]: Done.

File: include/gpgmm_d3d12.h
Comment (bjjones) [reply include/gpgmm_d3d12.h]: Done.

File: src/gpgmm/d3d12/ResourceHeapAllocatorD3D12.h
Comment (bbernhar) [inline src/gpgmm/d3d12/ResourceHeapAllocatorD3D12.h:26]: unit: rename to `CreateResourceHeapCallbackContext` to be consistent?

**Comment by bbernhar** [comment]
Dawn integration patch will also need to be updated too. Just apply the https://github.com/intel/GPGMM/blob/main/.github/workflows/.patches/dawn.diff to a local (Dawn) repo, modify it, then save the new diff. I keep the Dawn bot going for API coverage.

**Comment by bjjones** [comment]
Is the dawn.patch supposed to apply cleanly to top-of-tree Dawn? I'm getting multiple conflicts and even more errors when I try to build.

**Comment by bbernhar** [comment]
> Is the dawn.patch supposed to apply cleanly to top-of-tree Dawn? I'm getting multiple conflicts and even more errors when I try to build.

Possibly but it's currently based on https://github.com/intel/GPGMM/blob/main/scripts/dawn.deps (`be4c9f48aaef9f2144a654a847b76a6a0050ec5b`).

File: include/gpgmm_d3d12.h
Comment (bbernhar) [inline include/gpgmm_d3d12.h:184]: Leaving a TODO to update README.md as well.

File: include/gpgmm_d3d12.h
Comment (bbernhar) [inline include/gpgmm_d3d12.h:242]: Alternatively, we could have a CreateHeap2 (and means to not include std::function).

**Review by bbernhar — APPROVED** [review]
LGTM + after fixes.

File: src/tests/end2end/D3D12ResidencyManagerTests.cpp
Comment (bbernhar) [inline src/tests/end2end/D3D12ResidencyManagerTests.cpp:242]: nit: could we we not just pass `&wrapper` directly?

File: include/gpgmm_d3d12.h
Comment (bjjones) [reply include/gpgmm_d3d12.h:242]: If I understand correctly, you mean to leave the old CreateHeap that can accept a standard callback function without a context class? Do you think that would be useful? You wouldn't be able to access any variables (i.e. the D3D device).

File: src/tests/end2end/D3D12ResidencyManagerTests.cpp
Comment (bjjones) [reply src/tests/end2end/D3D12ResidencyManagerTests.cpp]: If by directly you mean without the cast, yes. Fixed.

File: src/mvi/gpgmm_d3d12.h
Comment (bjjones) [reply src/mvi/gpgmm_d3d12.h]: I just named them based off the function name that's calling CreateHeap. Doesn't matter to me, so I changed it.

File: include/gpgmm_d3d12.h
Comment (bjjones) [reply include/gpgmm_d3d12.h:184]: Updated with the new CreateHeap signature. It might be a little confusing without looking at the explanation in the header. Happy to add more detail if you like.

File: src/gpgmm/d3d12/ResourceHeapAllocatorD3D12.h
Comment (bjjones) [reply src/gpgmm/d3d12/ResourceHeapAllocatorD3D12.h]: Done.

**Comment by bbernhar** [comment]
> I'm getting hung up on updating the dawn.diff. It applies cleanly on the version you've stated, but I then I get a number of build errors. I also need to be using near top of tree GPGMM if I want to apply this CreateHeap change - and the build errors only get worse after that.

Normally I just edit + commit diff and let the builder do its thing. Otherwise, the steps to build locally are here: https://github.com/intel/GPGMM/blob/main/.github/workflows/win_dawn_rel.yaml. Typically, a `git checkout ` followed by `gclient sync` is all I need.

File: include/gpgmm_d3d12.h
Comment (bbernhar) [reply include/gpgmm_d3d12.h:242]: Yea, basically. As you can probably tell, it's cumbersome to not use a `std::function` (STL-compatible) for this. Not only does it break the API, it's harder to use. So a `CreateHeap2` entry point (mainly just for Chromium) might unblock you faster.

File: src/tests/end2end/D3D12ResidencyManagerTests.cpp
Comment (bbernhar) [inline src/tests/end2end/D3D12ResidencyManagerTests.cpp:79]: Could we please re-use the same (internal) implementation here (and elsewhere) and just set the flags it needs?

File: src/tests/capture_replay_tests/D3D12EventTraceReplay.cpp
Comment (bjjones) [reply src/tests/capture_replay_tests/D3D12EventTraceReplay.cpp]: Done.

File: src/tests/end2end/D3D12ResidencyManagerTests.cpp
Comment (bjjones) [reply src/tests/end2end/D3D12ResidencyManagerTests.cpp]: Done.

File: src/tests/end2end/D3D12ResidencyManagerTests.cpp
Comment (bjjones) [reply src/tests/end2end/D3D12ResidencyManagerTests.cpp]: Done.

File: src/tests/end2end/D3D12ResidencyManagerTests.cpp
Comment (bjjones) [reply src/tests/end2end/D3D12ResidencyManagerTests.cpp]: Ack.

File: src/gpgmm/d3d12/ResourceAllocatorD3D12.cpp
Comment (bbernhar) [inline src/gpgmm/d3d12/ResourceAllocatorD3D12.cpp:1245]: nit: Nothing is being created. Rename to ImportResourceCallbackContext?

File: include/gpgmm_d3d12.h
Comment (bbernhar) [inline include/gpgmm_d3d12.h:185]: unit: should be `pContext`

File: src/tests/end2end/D3D12ResidencyManagerTests.cpp
Comment (bbernhar) [reply src/tests/end2end/D3D12ResidencyManagerTests.cpp]: Just please confirm this (+ feedback from the CL reviewers) before you merge. Thanks!

File: src/gpgmm/d3d12/ResourceAllocatorD3D12.cpp
Comment (bjjones) [reply src/gpgmm/d3d12/ResourceAllocatorD3D12.cpp:1473]: I'm not sure which name you're referring to. You suggested calling the class name ImportResourceCallbackContext above, and every static callback member function in this PR is called CreateHeapWrapper.

Comment (bbernhar) [reply src/gpgmm/d3d12/ResourceAllocatorD3D12.cpp:1473]: Classes in `src/tests/end2end/D3D12ResidencyManagerTests.cpp`.

File: include/gpgmm_d3d12.h
Comment (bjjones) [reply include/gpgmm_d3d12.h:185]: Done.

File: src/gpgmm/d3d12/ResourceAllocatorD3D12.cpp
Comment (bjjones) [reply src/gpgmm/d3d12/ResourceAllocatorD3D12.cpp]: All changed to use CallbackContext suffix

File: src/gpgmm/d3d12/ResourceAllocatorD3D12.cpp
Comment (bjjones) [reply src/gpgmm/d3d12/ResourceAllocatorD3D12.cpp]: Done.

File: src/tests/end2end/D3D12ResidencyManagerTests.cpp
Comment (bjjones) [reply src/tests/end2end/D3D12ResidencyManagerTests.cpp]: Other CL comments appear to just be documentation. Will open another PR for those.

I don't know why someone would need to use this with Chromium's callback functions as I anticipated they would just use the class I've created here. It doesn't look like anything is stopping you if you want to call this with their callback functions though.

Comment (bbernhar) [reply src/tests/end2end/D3D12ResidencyManagerTests.cpp]: SGTM.

**Comment by bbernhar** [comment]
@bjjones Thanks!

**Pull request merged**
